src/dragonball/src/dbs_pci/src/msi.rs

```diff
@@ -0,0 +1,759 @@
+        size
+    }
+
+    /// Handle read accesses to the PCI MSI capability structure.
+    pub fn read(&mut self, offset: u64, data: &mut [u8]) -> std::io::Result<()> {
+        let (msg_data_offset, addr_hi_offset, mask_bits_offset) = self.get_offset_info();
+
+        // Please be really careful to touch code below, you should have good understanding of
+        // PCI MSI capability.
+        match data.len() {
+            1 => match offset {
+                0 => data[0] = self.cap_id_next as u8,
+                1 => data[0] = (self.cap_id_next >> 8) as u8,
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() => {
+                    data[0] = self.mask_bits as u8;
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 1 => {
+                    data[0] = (self.mask_bits >> 8) as u8;
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 2 => {
+                    data[0] = (self.mask_bits >> 16) as u8;
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 3 => {
+                    data[0] = (self.mask_bits >> 24) as u8;
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 4 => {
+                    self.get_pending_state(data, 0);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 5 => {
+                    self.get_pending_state(data, 1);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 6 => {
+                    self.get_pending_state(data, 2);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 7 => {
+                    self.get_pending_state(data, 3);
+                }
+                _ => debug!("invalid offset"),
+            },
+            2 => match offset {
+                0 => LittleEndian::write_u16(data, self.cap_id_next),
+                MSI_MSG_CTL_OFFSET => LittleEndian::write_u16(data, self.msg_ctl),
+                x if x == msg_data_offset => {
+                    LittleEndian::write_u16(data, self.msg_data);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() => {
+                    LittleEndian::write_u16(data, self.mask_bits as u16);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 2 => {
+                    LittleEndian::write_u16(data, (self.mask_bits >> 16) as u16);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 4 => {
+                    self.get_pending_state(data, 0);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 6 => {
+                    self.get_pending_state(data, 2);
+                }
```

Comment:
Ditto.

Reply:
this is for reading bits according to pci spec. I have added a summarised comment above
"Support writing mask_bits, MSI control information, message data, message address."